File: java/com/google/copybara/folder/FolderDestinationReader.java
/*
 * Copyright (C) 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.folder;

import static com.google.common.collect.ImmutableList.toImmutableList;

import com.google.common.collect.ImmutableList;
import com.google.copybara.CheckoutPath;
import com.google.copybara.DestinationReader;
import com.google.copybara.config.SkylarkUtil;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.util.Glob;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.util.stream.Stream;
import net.starlark.java.eval.EvalException;

/** A {@link DestinationReader} for reading files from a {@link FolderDestination}. */
public class FolderDestinationReader extends DestinationReader {
  private final Path folderPath;
  private final Path workDir;

  public FolderDestinationReader(Path folderPath, Path workDir) {
    this.folderPath = folderPath;
    this.workDir = workDir;
  }

  @Override
  public String readFile(String path) throws RepoException {
    try {
      return Files.readString(folderPath.resolve(path));
    } catch (IOException e) {
      throw new RepoException(String.format("Unable to read file %s.", path), e);
    }
  }

  @Override
  public void copyDestinationFiles(Object globObj, Object path)
      throws RepoException, ValidationException, EvalException {
    CheckoutPath checkoutPath = SkylarkUtil.convertFromNoneable(path, null);
    Glob glob = Glob.wrapGlob(globObj, null);
    if (checkoutPath == null) {
      copyDestinationFilesToDirectory(glob, workDir);
    } else {
      copyDestinationFilesToDirectory(
          glob, checkoutPath.getCheckoutDir().resolve(checkoutPath.getPath()));
    }
  }

  @Override
  public void copyDestinationFilesToDirectory(Glob glob, Path directory)
      throws RepoException, ValidationException {
    PathMatcher pathMatcher = glob.relativeTo(folderPath);

    for (String root : glob.roots()) {
      try (Stream<Path> stream = Files.walk(folderPath.resolve(root))) {
        ImmutableList<Path> matchingPaths =
            stream
                .filter(Files::isRegularFile)
                .filter(pathMatcher::matches)
                .collect(toImmutableList());

        for (Path sourcePath : matchingPaths) {
          Path targetPath = directory.resolve(folderPath.relativize(sourcePath));
          Files.createDirectories(targetPath.getParent());
          Files.copy(sourcePath, targetPath);
        }
      } catch (IOException e) {
        throw new RepoException(String.format("Failed to copy files from %s.", folderPath), e);
      }
    }
  }

  @Override
  public boolean exists(String path) {
    return Files.exists(folderPath.resolve(path));
  }
}

File: src/Copybara.Core/Folder/FolderDestinationReader.cs
/*
 * Copyright (C) 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Copybara.Exceptions;
using Copybara.Util;
using StarlarkRt = Starlark.Eval.Starlark;

namespace Copybara.Folder;

/// <summary>A <see cref="DestinationReader"/> for reading files from a <see cref="FolderDestination"/>.</summary>
public class FolderDestinationReader : DestinationReader
{
    private readonly string _folderPath;
    private readonly string _workDir;

    public FolderDestinationReader(string folderPath, string workDir)
    {
        _folderPath = folderPath;
        _workDir = workDir;
    }

    public override string ReadFile(string path)
    {
        try
        {
            return File.ReadAllText(Path.Combine(_folderPath, path));
        }
        catch (IOException e)
        {
            throw new RepoException($"Unable to read file {path}.", e);
        }
    }

    public override void CopyDestinationFiles(object glob, object path)
    {
        CheckoutPath? checkoutPath =
            ReferenceEquals(path, StarlarkRt.None) || path is null ? null : (CheckoutPath)path;
        Glob resolvedGlob = Glob.WrapGlob(glob, null)!;
        if (checkoutPath == null)
        {
            CopyDestinationFilesToDirectory(resolvedGlob, _workDir);
        }
        else
        {
            CopyDestinationFilesToDirectory(
                resolvedGlob,
                PathOps.Resolve(checkoutPath.GetCheckoutDir(), checkoutPath.GetPath()));
        }
    }

    public override void CopyDestinationFilesToDirectory(Glob glob, string directory)
    {
        IPathMatcher pathMatcher = glob.RelativeTo(_folderPath);

        foreach (var root in glob.Roots())
        {
            string rootPath = string.IsNullOrEmpty(root) ? _folderPath : Path.Combine(_folderPath, root);
            if (!Directory.Exists(rootPath))
            {
                continue;
            }
            try
            {
                foreach (var sourcePath in Directory.EnumerateFiles(
                    rootPath, "*", SearchOption.AllDirectories))
                {
                    string fullSource = Path.GetFullPath(sourcePath);
                    if (!pathMatcher.Matches(fullSource))
                    {
                        continue;
                    }
                    string relative = Path.GetRelativePath(_folderPath, fullSource);
                    string targetPath = Path.Combine(directory, relative);
                    string? parent = Path.GetDirectoryName(targetPath);
                    if (parent != null)
                    {
                        Directory.CreateDirectory(parent);
                    }
                    File.Copy(fullSource, targetPath);
                }
            }
            catch (IOException e)
            {
                throw new RepoException($"Failed to copy files from {_folderPath}.", e);
            }
        }
    }

    public override bool Exists(string path) =>
        File.Exists(Path.Combine(_folderPath, path)) || Directory.Exists(Path.Combine(_folderPath, path));
}
